src/render/painter.js

```diff
@@ -276,140 +275,111 @@ class Painter {
         this.renderPass = 'translucent';
-        {
-            let sourceCache;
-            let coords = [];
-            let symbolCoords = [];
-
-            this.currentLayer = 0;
-
-            for (this.currentLayer; this.currentLayer < layerIds.length; this.currentLayer++) {
-                const layer = this.style._layers[layerIds[this.currentLayer]];
-
-                if (layer.source !== (sourceCache && sourceCache.id)) {
-                    sourceCache = this.style.sourceCaches[layer.source];
-                    coords = [];
-                    symbolCoords = [];
-
-                    if (sourceCache) {
-                        this.clearStencil();
-                        coords = sourceCache.getVisibleCoordinates(false);
-                        // For symbol layers in the translucent pass, we add extra tiles to
-                        // the renderable set for cross-tile symbol fading.
-                        // Symbol layers don't use tile clipping, so no need to render
-                        // separate clipping masks
-                        symbolCoords = sourceCache.getVisibleCoordinates(true);
-                        if (sourceCache.getSource().isTileClipped) {
-                            this._renderTileClippingMasks(coords);
-                        }
-                    }
-
-                    coords.reverse();
-                    symbolCoords.reverse();
-                }
 
-                this.renderLayer(this, (sourceCache: any), layer, layer.type === 'symbol' ? symbolCoords : coords);
+        for (this.currentLayer = 0, prevSourceId = null; this.currentLayer < layerIds.length; this.currentLayer++) {
+            const layer = this.style._layers[layerIds[this.currentLayer]];
+            const sourceCache = sourceCaches[layer.source];
+
+            // For symbol layers in the translucent pass, we add extra tiles to the renderable set
+            // for cross-tile symbol fading. Symbol layers don't use tile clipping, so no need to render
+            // separate clipping masks
+            const coords = (layer.type === 'symbol' ? coordsDescendingSymbol : coordsDescending)[layer.source];
+
+            if (layer.source !== prevSourceId && sourceCache) {
+                this.clearStencil();
+                if (sourceCache.getSource().isTileClipped) {
+                    this._renderTileClippingMasks(coordsAscending[layer.source]);
+                }
             }
+
+            this.renderLayer(this, sourceCache, layer, coords);
+            prevSourceId = layer.source;
         }
 
         if (this.options.showTileBoundaries) {
-            const sourceCache = this.style.sourceCaches[Object.keys(this.style.sourceCaches)[0]];
-            if (sourceCache) {
-                draw.debug(this, sourceCache, sourceCache.getVisibleCoordinates());
+            for (const id in sourceCaches) {
+                draw.debug(this, sourceCaches[id], coordsAscending[id]);
```

Comment:
It doesn't look like any behavior changes here, but it got me thinking -- would it be better/more consistent if we drew debug tile boundaries as the union of tiles in all source caches?

Reply:
We used to draw debug boundaries for all source caches, but I felt it looked very confusing with multiple sources. You can try it but I think it's more useful to have clear, legible boundaries for at least one source.